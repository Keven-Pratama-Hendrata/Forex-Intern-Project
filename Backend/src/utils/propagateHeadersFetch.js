import fetch from 'node-fetch';

export default async function propagateHeadersFetch(req, url, options = {}, propagateHeaders = ['x-flow-id']) {
    const outgoingHeaders = { ...options.headers };
    for (const header of propagateHeaders) {
        const value = req.header(header);
        if (value) {
            outgoingHeaders[header] = value;
        }
    }
    return fetch(url, { ...options, headers: outgoingHeaders });
} 