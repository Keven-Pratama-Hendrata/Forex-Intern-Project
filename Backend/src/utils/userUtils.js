import User from "../../models/User.js";

export async function updateUserBalance(userId, currency, amount) {
    const user = await User.findById(userId);
    if (!user) {
        throw new Error('User not found');
    }

    if (!user.balances.hasOwnProperty(currency)) {
        throw new Error('Invalid currency');
    }

    const newBalance = parseFloat(user.balances[currency]) + parseFloat(amount);
    if (newBalance < 0) {
        throw new Error('Insufficient funds');
    }

    user.balances[currency] = newBalance;
    user.balanceHistory.push({
        currency,
        balance: newBalance,
        amount: parseFloat(amount),
        date: new Date()
    });

    return user.save();
}

export async function findUserByUsername(username) {
    const user = await User.findOne({ userName: username });
    if (!user) {
        throw new Error('User not found');
    }
    return user;
}