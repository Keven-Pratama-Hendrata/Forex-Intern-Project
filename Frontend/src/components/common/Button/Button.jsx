import React from 'react';
import {
  getButtonEventHandlers,
  getButtonCoreProps,
  renderButton
} from './buttonUtils.jsx';

/**
 * Styled button component with interactive hover and press effects.
 * Renders a full-width, rounded button with custom color and transition effects.
 *
 * @param {Object} props Button properties.
 * @param {React.ReactNode} props.children Content to display inside the button (text or elements).
 * @param {string} [props.type='button'] The button type attribute (e.g., 'button', 'submit', 'reset').
 * @param {boolean} [props.disabled=false] If true, disables the button and prevents interaction.
 * @param {string} [props.className] Additional class names for custom styling.
 * @param {function} [props.onClick] Function to call when the button is clicked.
 * @returns {JSX.Element} A styled, interactive button for use in forms and UI.
 */
function Button({
  children,
  type = 'button',
  disabled = false,
  className = '',
  onClick,
  ...props
}) {
  const coreProps = getButtonCoreProps({ children, type, disabled, className, ...props });
  const eventHandlers = getButtonEventHandlers(onClick);
  return renderButton({ ...coreProps, ...eventHandlers });
}

export default Button; 