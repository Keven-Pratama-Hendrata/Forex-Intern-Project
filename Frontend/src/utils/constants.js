export const BLUE_BUTTON = "#507fa1";
export const BLUE_BUTTON_HOVER = "#426989"; 