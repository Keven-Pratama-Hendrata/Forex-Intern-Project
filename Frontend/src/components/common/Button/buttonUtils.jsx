import { BLUE_BUTTON, BLUE_BUTTON_HOVER } from '../../../utils/constants';

/**
 * Sets the button background color to the hover color on mouse enter.
 * @param {React.MouseEvent} e Mouse event.
 * @returns {void}
 */
export const handleMouseEnter = (e) => (e.currentTarget.style.backgroundColor = BLUE_BUTTON_HOVER);

/**
 * Resets the button background color on mouse leave.
 * @param {React.MouseEvent} e Mouse event.
 * @returns {void}
 */
export const handleMouseLeave = (e) => (e.currentTarget.style.backgroundColor = BLUE_BUTTON);

/**
 * Scales the button down on mouse down.
 * @param {React.MouseEvent} e Mouse event.
 * @returns {void}
 */
export const handleMouseDown = (e) => (e.currentTarget.style.transform = 'scale(0.97)');

/**
 * Resets the button scale on mouse up.
 * @param {React.MouseEvent} e Mouse event.
 * @returns {void}
 */
export const handleMouseUp = (e) => (e.currentTarget.style.transform = 'scale(1)');

/**
 * Returns the full className string for the button.
 * @param {string} className Additional class names.
 * @returns {string} The full className string.
 */
export function getButtonClassName(className) {
    return [
        'btn w-full rounded-full border-none text-white font-semibold',
        'tracking-wide transition-transform duration-200',
        className
    ].join(' ');
}

/**
 * Returns the style object for the button.
 * @returns {Object} The style object.
 */
export function getButtonStyle() {
    return { backgroundColor: BLUE_BUTTON };
}

/**
 * Returns event handlers for button interactions.
 * @param {function} onClick Click handler.
 * @returns {Object} Event handlers for the button.
 */
export function getButtonEventHandlers(onClick) {
    return {
        onMouseEnter: handleMouseEnter,
        onMouseLeave: handleMouseLeave,
        onMouseDown: handleMouseDown,
        onMouseUp: handleMouseUp,
        onClick,
    };
}

/**
 * Returns the core props for the button element.
 * @param {Object} props Button props.
 * @param {React.ReactNode} props.children Button content.
 * @param {string} [props.type] Button type.
 * @param {boolean} [props.disabled] Disabled state.
 * @param {string} [props.className] Additional class names.
 * @returns {Object} The core props for the button.
 */
export function getButtonCoreProps({ children, type, disabled, className, ...props }) {
    return {
        type,
        disabled,
        className: getButtonClassName(className),
        style: getButtonStyle(),
        ...props,
        children,
    };
}

/**
 * Renders the button element with the given props.
 * @param {Object} props Button props.
 * @returns {JSX.Element} The button element.
 */
export function renderButton(props) {
    return (
        <button {...props}>{props.children}</button>
    );
} 