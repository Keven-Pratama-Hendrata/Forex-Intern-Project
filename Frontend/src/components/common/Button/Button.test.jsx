import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { describe, it, expect, jest } from '@jest/globals';

import Button from './Button';
import { BLUE_BUTTON, BLUE_BUTTON_HOVER } from '../../../utils/constants';

describe('Button', () => {
  it('renders with children', () => {
    render(<Button>Click me</Button>);
    expect(screen.getByRole('button')).toHaveTextContent('Click me');
  });

  it('applies custom className and type', () => {
    render(
      <Button type="submit" className="my-class" data-testid="custom-btn">X</Button>
    );
    const btn = screen.getByTestId('custom-btn');
    expect(btn).toHaveAttribute('type', 'submit');
    expect(btn).toHaveClass('my-class');
  });

  it('is disabled when disabled prop is true', () => {
    render(<Button disabled>Disabled</Button>);
    expect(screen.getByRole('button')).toBeDisabled();
  });

  it('fires onClick when enabled', () => {
    const handler = jest.fn();
    render(<Button onClick={handler}>OK</Button>);
    fireEvent.click(screen.getByRole('button'));
    expect(handler).toHaveBeenCalledTimes(1);
  });

  it('does not fire onClick when disabled', () => {
    const handler = jest.fn();
    render(<Button onClick={handler} disabled>NO</Button>);
    fireEvent.click(screen.getByRole('button'));
    expect(handler).not.toHaveBeenCalled();
  });

  it('applies interactive styles on hover and press', () => {
    render(<Button>Test</Button>);
    const btn = screen.getByRole('button');
    expect(btn).toHaveStyle(`background-color: ${BLUE_BUTTON}`);
    fireEvent.mouseEnter(btn);
    expect(btn).toHaveStyle(`background-color: ${BLUE_BUTTON_HOVER}`);
    fireEvent.mouseDown(btn);
    expect(btn).toHaveStyle('transform: scale(0.97)');
    fireEvent.mouseUp(btn);
    expect(btn).toHaveStyle('transform: scale(1)');
    fireEvent.mouseLeave(btn);
    expect(btn).toHaveStyle(`background-color: ${BLUE_BUTTON}`);
  });

  it('matches snapshot', () => {
    const { asFragment } = render(
      <Button type="button" className="snapshot-btn" disabled={false} onClick={() => { }}>
        Snapshot
      </Button>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});