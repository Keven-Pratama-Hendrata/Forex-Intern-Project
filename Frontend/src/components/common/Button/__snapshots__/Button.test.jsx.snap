// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Button matches snapshot 1`] = `
<DocumentFragment>
  <button
    class="btn w-full rounded-full border-none text-white font-semibold tracking-wide transition-transform duration-200 snapshot-btn"
    style="background-color: rgb(80, 127, 161);"
    type="button"
  >
    Snapshot
  </button>
</DocumentFragment>
`;
